File: python/librats_py/examples/file_transfer.py
#!/usr/bin/env python3
"""
File transfer example using librats.

Demonstrates the push-model file-transfer subsystem: one peer offers a file,
the other accepts it. The file-transfer subsystem and its callbacks must be
enabled/registered before ``start()``.
"""

import os
import sys

from librats_py import RatsNode, RatsError, LogLevel, FileTransferStatus, set_log_level


class FileTransferExample:
    def __init__(self, listen_port: int, download_dir: str):
        self.listen_port = listen_port
        self.download_dir = download_dir
        os.makedirs(download_dir, exist_ok=True)

        self.node = RatsNode(listen_port)

        # Register callbacks + enable subsystem BEFORE start().
        self.node.on_peer_connected(lambda pid: print(f"+ Peer connected: {pid}"))
        self.node.on_peer_disconnected(
            lambda pid, why: print(f"- Peer disconnected: {pid} ({why})"))

        self.node.enable_file_transfer(temp_dir=download_dir)
        self.node.on_file_offer(self.on_offer)
        self.node.on_file_progress(self.on_progress)
        self.node.on_file_complete(self.on_complete)

    def on_offer(self, peer_id, transfer_id, name, size, is_directory):
        kind = "directory" if is_directory else "file"
        print(f"\nIncoming {kind} offer '{name}' ({size} bytes) "
              f"from {peer_id[:16]}… (transfer {transfer_id})")
        dest = os.path.join(self.download_dir, name)
        # Auto-accept for the demo.
        self.node.accept_file(peer_id, transfer_id, dest)
        print(f"Accepted → {dest}")

    def on_progress(self, transfer_id, peer_id, done, total, status):
        pct = (done * 100 // total) if total else 0
        print(f"  transfer {transfer_id}: {pct}% ({done}/{total}) "
              f"[{FileTransferStatus(status).name}]")

    def on_complete(self, transfer_id, success, path):
        state = "completed" if success else "failed"
        print(f"\nTransfer {transfer_id} {state}: {path}")

    def start(self):
        self.node.start()
        print(f"File transfer client on port {self.listen_port}")
        print(f"Local peer id: {self.node.local_id}")

    def stop(self):
        self.node.stop()

    def run(self):
        print("\nCommands:")
        print("  connect <host> <port>        - dial a peer")
        print("  send <peer_id> <file_path>   - offer a file")
        print("  senddir <peer_id> <dir_path> - offer a directory")
        print("  peers                        - list peers")
        print("  quit                         - exit")
        print()
        while True:
            try:
                parts = input("file-transfer> ").strip().split()
            except (EOFError, KeyboardInterrupt):
                break
            if not parts:
                continue
            cmd = parts[0].lower()
            try:
                if cmd in ("quit", "exit"):
                    break
                elif cmd == "connect" and len(parts) == 3:
                    self.node.connect(parts[1], int(parts[2]))
                    print(f"Dialing {parts[1]}:{parts[2]}…")
                elif cmd == "send" and len(parts) == 3:
                    if not os.path.exists(parts[2]):
                        print("File not found.")
                        continue
                    tid = self.node.send_file(parts[1], parts[2])
                    print(f"Offered file (transfer {tid})")
                elif cmd == "senddir" and len(parts) == 3:
                    tid = self.node.send_directory(parts[1], parts[2])
                    print(f"Offered directory (transfer {tid})")
                elif cmd == "peers":
                    for pid in self.node.peer_ids:
                        print(f"  - {pid}")
                else:
                    print("Unknown/invalid command.")
            except RatsError as e:
                print(f"Error: {e}")


def main():
    if len(sys.argv) < 2:
        print("Usage: python file_transfer.py <listen_port> [download_dir]")
        sys.exit(1)
    try:
        listen_port = int(sys.argv[1])
    except ValueError:
        print("Error: Port must be a number")
        sys.exit(1)
    download_dir = sys.argv[2] if len(sys.argv) > 2 else "./downloads"

    set_log_level(LogLevel.INFO)
    example = FileTransferExample(listen_port, download_dir)
    try:
        example.start()
        example.run()
    finally:
        example.stop()


if __name__ == "__main__":
    main()

File: nodejs/examples/file_transfer.js
/**
 * File transfer (push model).
 *
 * The sender offers with sendFile()/sendDirectory(); the receiver gets an
 * onFileOffer() and answers with acceptFile()/rejectFile() — the
 * (peerId, transferId) pair names the transfer on both sides. Progress and
 * completion arrive on onFileProgress()/onFileComplete().
 *
 * The subsystem and all three callbacks are registered BEFORE start().
 *
 *   node file_transfer.js [listenPort] [connectHost] [connectPort]
 */

const fs = require('fs');
const path = require('path');
const readline = require('readline');
const { RatsNode } = require('../lib/index');

const [portArg, host, peerPortArg] = process.argv.slice(2);

const downloadDir = path.join(__dirname, 'transfers', 'downloads');
const tempDir = path.join(__dirname, 'transfers', 'temp');
fs.mkdirSync(downloadDir, { recursive: true });

const node = new RatsNode(portArg ? Number(portArg) : 8080);

node.onPeerConnected((peerId) => console.log(`+ peer ${peerId}`));
node.onPeerDisconnected((peerId, reason) => console.log(`- peer ${peerId} (${reason})`));

// tempDir holds in-progress downloads; they move to their destination on completion.
node.enableFileTransfer(tempDir);

// Auto-accept every offer into downloadDir. A real app would ask the user here.
node.onFileOffer((peerId, transferId, name, size, isDirectory) => {
  const dest = path.join(downloadDir, name);
  console.log(
    `offer [${transferId}] ${isDirectory ? 'dir' : 'file'} "${name}" ` +
    `(${size} bytes) from ${peerId} -> ${dest}`
  );
  node.acceptFile(peerId, transferId, dest);
});

node.onFileProgress((transferId, peerId, done, total, status) => {
  const pct = total > 0 ? Math.round((done / total) * 100) : 0;
  console.log(`[${transferId}] ${pct}% (${done}/${total}) status=${status}`);
});

node.onFileComplete((transferId, peerId, success, filePath) => {
  console.log(success ? `[${transferId}] done: ${filePath}` : `[${transferId}] failed`);
});

node.start();

console.log(`peer id   ${node.localId}`);
console.log(`downloads ${downloadDir}`);

if (host) node.connect(host, Number(peerPortArg) || 8081);

// ---- interactive prompt ----

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  prompt: 'librats> ',
});

console.log([
  '',
  'commands:',
  '  connect <host> <port>          dial a peer',
  '  send <peerId> <file>           offer a file',
  '  senddir <peerId> <dir>         offer a directory tree',
  '  pause|resume|cancel <peerId> <transferId>',
  '  peers                          list connected peers',
  '  quit                           exit',
  '',
].join('\n'));

rl.prompt();

rl.on('line', (line) => {
  const [command, ...rest] = line.trim().split(/\s+/);
  try {
    switch (command) {
      case '':
        break;
      case 'connect':
        need(rest, 2, 'connect <host> <port>') && node.connect(rest[0], Number(rest[1]));
        break;
      case 'send':
        need(rest, 2, 'send <peerId> <file>') && offer('sendFile', rest[0], rest[1]);
        break;
      case 'senddir':
        need(rest, 2, 'senddir <peerId> <dir>') && offer('sendDirectory', rest[0], rest[1]);
        break;
      case 'pause':
      case 'resume':
      case 'cancel':
        need(rest, 2, `${command} <peerId> <transferId>`) &&
          node[`${command}File`](rest[0], Number(rest[1]));
        break;
      case 'peers':
        console.log(node.peerIds.join('\n') || '(none)');
        break;
      case 'quit':
      case 'exit':
        rl.close();
        return;
      default:
        console.log(`unknown command: ${command}`);
    }
  } catch (err) {
    console.error(err.message);
  }
  rl.prompt();
});

rl.on('close', () => {
  node.stop();
  process.exit(0);
});

function need(args, count, usage) {
  if (args.length >= count) return true;
  console.log(`usage: ${usage}`);
  return false;
}

function offer(method, peerId, target) {
  if (!fs.existsSync(target)) {
    console.log(`not found: ${target}`);
    return;
  }
  // 0 means the offer was refused outright — subsystem off, or unknown peer.
  const transferId = node[method](peerId, target);
  console.log(transferId ? `offering "${target}" as [${transferId}]` : 'offer refused');
}
